store/localstore/kv.go

```diff
@@ -29,16 +28,224 @@ var (
 	_ kv.Storage = (*dbStore)(nil)
 )
 
+type op int
+
+const (
+	opSeek = iota + 1
+	opCommit
+)
+
+type command struct {
+	op    op
+	txn   *dbTxn
+	args  interface{}
+	reply interface{}
+	done  chan error
+}
+
+type seekReply struct {
+	key   []byte
+	value []byte
+	err   error
+}
+
+type commitReply struct {
+	err error
+}
+
+type seekArgs struct {
+	key []byte
+}
+
+type commitArgs struct {
+}
+
+//scheduler
+
+// Seek searches for the first key in the engine which is >= key in byte order, returns (nil, nil, ErrNotFound)
+// if such key is not found.
+func (s *dbStore) Seek(key []byte) ([]byte, []byte, error) {
+	c := &command{
+		op:   opSeek,
+		args: &seekArgs{key: key},
+		done: make(chan error, 1),
+	}
+
+	s.commandCh <- c
+	err := <-c.done
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	reply := c.reply.(*seekReply)
+	return reply.key, reply.value, nil
+}
+
+// Commit writes the changed data in Batch.
+func (s *dbStore) CommitTxn(txn *dbTxn) error {
+	if len(txn.snapshotVals) == 0 {
+		return nil
+	}
+	c := &command{
+		op:   opCommit,
+		txn:  txn,
+		args: &commitArgs{},
+		done: make(chan error, 1),
+	}
+
+	s.commandCh <- c
+	err := <-c.done
+	return errors.Trace(err)
+}
+
+func (s *dbStore) seekWorker(seekCh chan *command) {
+	for {
+		var pending []*command
+		select {
+		case cmd, ok := <-seekCh:
+			if !ok {
+				return
+			}
+			pending = append(pending, cmd)
+		L:
+			for {
+				select {
+				case cmd, ok := <-seekCh:
+					if !ok {
+						break L
+					}
+					pending = append(pending, cmd)
+				default:
+					break L
+				}
+			}
+		}
+
+		s.doSeek(pending)
+	}
+}
+
+func (s *dbStore) scheduler() {
+	closed := false
+	seekCh := make(chan *command, 1000)
+	for i := 0; i < 3; i++ {
+		go s.seekWorker(seekCh)
+	}
+
+	for {
+		select {
+		case cmd := <-s.commandCh:
+			if closed {
+				cmd.done <- ErrDBClosed
+				continue
+			}
+			switch cmd.op {
+			case opSeek:
+				seekCh <- cmd
+			case opCommit:
+				s.doCommit(cmd)
+			}
+		case <-s.closeCh:
+			closed = true
+			s.wg.Done()
+			// notify seek worker to exit
+			close(seekCh)
+		}
+	}
+}
+
+func (s *dbStore) tryLock(txn *dbTxn) (err error) {
+	// check conflict
+	for k := range txn.snapshotVals {
+		if _, ok := s.keysLocked[k]; ok {
+			return errors.Trace(kv.ErrLockConflict)
+		}
+
+		lastVer, ok := s.recentUpdates[k]
+		if !ok {
+			continue
+		}
+		// If there's newer version of this key, returns error.
+		if lastVer.Cmp(kv.Version{Ver: txn.tid}) > 0 {
+			return errors.Trace(kv.ErrConditionNotMatch)
+		}
+	}
+
+	// record
+	for k := range txn.snapshotVals {
```

Comment:
There are two phases logically. We need to roll back if we record it then we got conflict.